StyleProperties::fontValue needs to return emptyString in many more cases where font sub properties have unusual values

darinadler opened the pull request:
#### d3d6e16bf1f0bab4f22043c2c8e326c9a51e315c
<pre>
StyleProperties::fontValue needs to return emptyString in many more cases where font sub properties have unusual values
<a href="https://bugs.webkit.org/show_bug.cgi?id=247498">https://bugs.webkit.org/show_bug.cgi?id=247498</a>
rdar://problem/102248342

Reviewed by Tim Nguyen and Sam Weinig.

To make the font serialization work correctly, had to change font shorthand parsing to set font
subproperties to initial values, not to the keyword &quot;initial&quot;. To make system font serialization
work correctly, had to change font shorthand parsing to set the system font for all longhands,
not just a select few.

* LayoutTests/fast/css/font-shorthand-expected.txt: Rebased since we use initial values, not
values that are literally &quot;initial&quot;, for the font subproperties set by the font shorthand.

* LayoutTests/TestExpectations: Removed expectations for tests that were already removed.

* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/font-shorthand-serialization-prevention-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/font-shorthand-serialization-prevention.html:
Added.

* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-shorthand-variant-expected.txt:
Expect PASS.

* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/system-fonts-serialization.tentative-expected.txt:
Expect PASS.

* LayoutTests/imported/w3c/web-platform-tests/css/cssom/font-shorthand-serialization-expected.txt:
Rebased since we use initial values, not values that are literally &quot;initial&quot;, for the font
subproperties set by the font shorthand.
* LayoutTests/platform/win/fast/css/font-shorthand-expected.txt: Ditto.

* LayoutTests/platform/win/imported/w3c/web-platform-tests/css/css-fonts/font-shorthand-serialization-prevention-expected.txt:
Added.

* Source/WebCore/css/CSSProperties.json: Removed the converter codegen property from
font-variant-caps, since the generator now handles it correctly.

* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::fontShorthandValue): Added checking for font-kerning, and removed the FIXME.

* Source/WebCore/css/StyleProperties.cpp:
(WebCore::valueID): Added.
(WebCore::StyleProperties::propertyAsValueID const): Changed to return a std::optional so
callers can tell a missing value from a non-valueID value; useful in more places.
(WebCore::StyleProperties::isSingleFontShorthand const): Deleted.
(WebCore::StyleProperties::fontValue const): Combined the CSS-wide keyword logic with the system
font keyword logic, and make sure we don&apos;t ever try to use either as an individual value in the
font shorthand. Added checking for the font subproperties that are reset by the font shorthand.
If any are set to non-initial values, then serialize as the empty string.
(WebCore::StyleProperties::fontSynthesisValue const): Removed valueID lambda since we can use the
function added above instead.
(WebCore::StyleProperties::getPropertyCSSValue const): Simplify a bit using the valueID function.

* Source/WebCore/css/StyleProperties.h: Changed return type of propertyAsValueID to std::optional.
Removed isSingleFontShorthand.

* Source/WebCore/css/StylePropertyShorthand.h: Added array for handling the font shorthand rules
for resetting font subproperties to their initial values, named
fontShorthandSubpropertiesResetToInitialValues. Used in two places. Maybe generate in future?

* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::CSSPropertyParser::consumeFont): Rewrote system font handling so it puts the system
font value into all font subproperties, not just a select few. Changed property resetting so
it resets properties to their initial values, not to the &quot;initial&quot; keyword.

* Source/WebCore/css/process-css-properties.py:
(GenerationContext): Generate code to apply the initial value when we encounter a system font.

* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertFontFeatureSettings): Handle system font.
(WebCore::Style::BuilderConverter::convertFontVariantCaps): Deleted. This is now generated.
(WebCore::Style::BuilderConverter::convertFontVariationSettings): Handle system font.
(WebCore::Style::BuilderConverter::convertFontPalette): Ditto.

* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyValueLineHeight): Handle system font.
(WebCore::Style::BuilderCustom::applyValueFontVariantLigatures): Ditto.
(WebCore::Style::BuilderCustom::applyValueFontVariantNumeric): Ditto.
(WebCore::Style::BuilderCustom::applyValueFontVariantEastAsian): Ditto.
(WebCore::Style::BuilderCustom::applyValueFontVariantAlternates): Ditto.
(WebCore::Style::BuilderCustom::applyValueFontSizeAdjust): Ditto.

* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::getPlatformEditorStateCommon const): Updated for change to propertyAsValueID.

Canonical link: <a href="https://commits.webkit.org/256635@main">https://commits.webkit.org/256635@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c5fc0f8009e2fe6c52ef505ce06c944fb421f73e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/96396 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/5650 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/29461 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/105939 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/166286 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/100379 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/5821 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/34396 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/88775 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/102665 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/102067 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/4334 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/82997 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/31317 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/86176 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/88065 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/74208 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/40126 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/19558 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/37800 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/20945 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4603 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/3091 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/43516 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/706 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/40214 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -575,7 +576,87 @@ static std::optional<CSSValueID> fontStretchKeyword(double value)
+        if (int foundPropertyIndex = findPropertyIndex(property); foundPropertyIndex != -1)
+            return propertyReferenceKeywordValue(propertyAt(foundPropertyIndex));
+        return std::nullopt;
+    };
```
My gut feeling is that there is prior art for these (or similar versions) in StyleProperties. Have you checked?

nt1m replied:
Also could the fontStretchPercentageAsKeyword logic benefit from any of these new helpers?

darinadler replied:
No, these are unique. There is a function for comparing with a specific value ID, but that’s not the same thing. I’d be open to more refactoring and reorganizing to be cleaner, but what you are seeing here is after multiple rounds of tightening and looking around for helper functions.

I am noticing that `propertyReferenceKeywordValue` in particular is not something that needs to capture anything. Could just be a helper function rather than a lambda.

But the `propertyKeywordValue` function is something that maybe _should_ have been elsewhere in this file. Instead other places use wordy code with lots of local variables for this sort of thing.

nt1m commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -575,7 +576,87 @@ static std::optional<CSSValueID> fontStretchKeyword(double value)
+        return isCSSWideKeyword(propertyReferenceKeywordValue(property));
+    };
+
+    auto propertiesResetByShorthandAreExpressible = [&] {
```
Could the font-stretch check go in here as well?

darinadler replied:
The `font-stretch` check happens as a side effect of mapping to a keyword. If I separated the check from the keyword mapping it would be less efficient, so I left it as is for now.

nt1m commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -609,24 +690,21 @@ String StyleProperties::fontValue() const
 
-    auto fontSizeString = fontSizeProperty.value()->cssText();
-    auto commonValue = fontSizeString;
+    // This code no longer uses commonValue, but we define it so we can use appendFontLonghandValueIfExplicit.
```
I find this a bit unfortunate, even if we save a bit of string builder work, since we now have the set of property names repeated 3 times.

darinadler replied:
I think that we can and will work on making this better over time. The special considerations for various longhands makes it hard to drive this from a single list of properties, but I won’t give up on that. For example, when actually serializing `font-stretch` needs lots of special handling, but the rules about CSS-wide keywords are the same as for every other property.

It’s a bad pattern to first call `cssText` and then after the fact check if the string happens to be a CSS-wide keyword. We should get rid of that completely. But it’s also bad to look up the same property more than once.

We need to do some of these same kinds of checks for `fontVariantValue`, so then the `appendFontLonghandValueIfExplicit` function can be changed too since that is the only other function that uses it.

It would be nice if this function could share lists of properties with the `consumeFont` function. I tried to take a step in that direction when rewriting `StyleProperties::isSingleFontShorthand` and it made the function much shorter too!

darinadler replied:
The whole `commonValue` idiom doesn’t really work, and is based on false ideas about the task. Non-CSS-wide-keyword values shouldn’t be checked for "common" values, it’s just a waste. And keywords like `inherit` can’t appear as values for single items within the shorthand, so they need to be handled differently whether they are "common" or not. This is handled wrong in the other shorthands, and is part of what we need to fix.

nt1m commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -575,7 +576,87 @@ static std::optional<CSSValueID> fontStretchKeyword(double value)
+            || propertyIsNonInitialCSSWideKeyword(CSSPropertyFontWeight)
+            || propertyIsNonInitialCSSWideKeyword(CSSPropertyFontStretch)
+            || propertyIsNonInitialCSSWideKeyword(CSSPropertyLineHeight);
+    };
```
nit: I tend to define lambdas right before I use them (e.g. before the call to `anyOptionalLonghandIsNonInitialCSSWideKeyword`), reduces a bit of cognitive overhead when reading :)

darinadler replied:
I’ve moved the lambdas around many times while working on this. It turned out the exact sequence of operations was important, and interspersing the lambdas with the code obscured some of the fine points of why things were done in the order they are. That’s why I ended up like this.

But I will change it to follow the "define lambdas right before I use them" pattern; that’s OK with me.

nt1m reviewed (commented):
Looks good in general, though I wonder if the logic could be shortened/simplified.

nt1m commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -105,6 +106,21 @@ static bool isValueIDIncludingList(const RefPtr<CSSValue>& value, CSSValueID id)
+static bool isCSSWideKeyword(const CSSValue& value)
+{
+    return isCSSWideKeyword(valueID(&value));
+}
```
Not sure this is worth adding, given there's `CSSValue::isCSSWideKeyword()`, so `value.isCSSWideKeyword()` would just work (this is also what `fontSynthesisValue()` uses).

darinadler replied:
Good suggestion.

nt1m commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -575,20 +547,79 @@ static std::optional<CSSValueID> fontStretchKeyword(double value)
+            return emptyString();
+    }
+
+    auto anyOptionalLonghandIsNonInitialCSSWideKeyword = [&] {
```
We shouldn't exclude `initial`:

`font: initial 16px Arial` or `font: 16px/initial Arial` are invalid serializations. Seems to be a bug in pre-existing code.

Testcase:
```
div.style = "font-family: Arial; font-size: 16px; font-style: initial;"
console.log(div.style.font)
```

This gives an empty string in Gecko/Blink, and currently gives `"initial 16px Arial"` in WebKit (which is an invalid value at parse-time).

Seems similar to https://bugs.webkit.org/show_bug.cgi?id=242775, perhaps a more general solution should be found at some point.

darinadler replied:
I will add a test case and make sure I have fixed that. I also noticed this pre-existing bug, which I think exists for many other short hands as well not just `font`.

However, it won’t be as simple as **just** removing the check for `initial`. If I do that alone, tests start failing.

darinadler replied:
I think I may have to rewrite the part of `consumeFont` that sets properties to default values. I was doing it with the initial keyword, but it seems that is not correct.

darinadler replied:
OK, dealt with this in the new patch. Had to do a lot!

nt1m reviewed (approved):
Looks good, thanks!

weinig commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -609,24 +640,21 @@ String StyleProperties::fontValue() const
 
-    auto fontSizeString = fontSizeProperty.value()->cssText();
-    auto commonValue = fontSizeString;
+    // This code no longer uses commonValue, for now we define it so we can use appendFontLonghandValueIfExplicit.
```
Looks like there are enough callers that don't actually care about the commonValue parameter that it might be nice to add an overload of appendFontLonghandValueIfExplicit that doesn't take one (and either re-implement or just call passing a dummy common value like here).

darinadler replied:
Yes, I expect to make that change, soon, but not in this pull request. The `appendFontLonghandValueIfExplicit` function is only used in two places, `fontValue` and `fontVariantValue`, and the latter has a number of bugs similar to the ones I am fixing here in `fontValue`. I suspect this function will be revised or eliminated entirely when I do that work.

weinig commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -488,78 +506,32 @@ void StyleProperties::appendFontLonghandValueIfExplicit(CSSPropertyID propertyID
-    if (sizeValueID == CSSValueInvalid)
-        return std::nullopt;
+    // This is different from the list of longhands in fontShorthand().
+    // This is the set of longhands that are *set* by the font shorthand, not including the many others that are reset to their initial values.
```
Is this kind of set something we need for any other shorthands? If so, we consider adding it to the data in CSSProperties.json, and producing some constexpr data / functions for it.

darinadler replied:
The font shorthand has a special consideration, where it resets the values of many longhands to initial values that can’t even be specified in the font shorthand. So I’m not sure how much there will be a repeating pattern.

One thing I did differently in the revised version of this pull request I am currently preparing is that I put this list of properties and values into a header because it’s now used in two places. It would be fine to later refactor so it is generated; would not be a big change.

It’s possible we may need to know the initial values for other shorthands; I am not sure.

weinig commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -575,20 +547,79 @@ static std::optional<CSSValueID> fontStretchKeyword(double value)
 
 String StyleProperties::fontValue() const
```
I am starting to think that maybe having a StylePropertiesSerialization (or whatnot) type to encapsulate and/or federate out the serialization logic might be a nice cleanup at some point as this file is a bit huge. (I also think we should split MutableStyleProperties and ImmutableStyleProperties into their own files).

darinadler replied:
I agree with you. Further, there are three different things that are done with various properties, parsing, creation of computed style, and special cases for serialization of shorthands. All three have to be kept in sync for a shorthand like `font`. Better organization of this could improve things quite a bit. It never occurred to me until now, but it’s strange that for computed style we actually produce a `CSSValue` for `font`, but for specified style we build a string without ever creating a `CSSValue`.

darinadler commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -5328,6 +5328,22 @@ bool CSSPropertyParser::parseFontPaletteValuesDescriptor(CSSPropertyID propId)
+        // Instead, we store sentinel values, later replaced by environment-sensitive values
+        // inside Style::BuilderCustom and Style::BuilderConverter.
+        for (auto property : fontShorthand())
+            addProperty(property, CSSPropertyFont, CSSValuePool::singleton().createIdentifierValue(systemFont), important, true);
```
Should probably remove the `true` here since that marks these as "implicit", whatever that means.

nt1m replied:
The original point  was to represent whether values were explicit set or not, to include/exclude them from serialization, but the CSSOM spec has changed since, and now serialization should be the same regardless of whether a property was implicitly or explicitly set. So now it's mostly being used as a quick hack to makes longhands appear or disappear from serialization (because it's much faster than creating a custom function in StyleProperties). I think we should just get rid of that concept in the future for that reason.

As for now, it's fine to use it however is convenient to you.

darinadler replied:
OK I will leave this alone for now.

nt1m commented on Source/WebCore/css/CSSProperties.json:
```diff
@@ -116,11 +116,13 @@
         "",
+        "* font-subproperty:",
+        "Indicates that this CSS property is a subproperty of the font shorthand.",
+        "",
```
I have a slight preference to not add new fields (given there's quite a bit of them) and to use the list of longhands from `font`. It's also unclear how/why it differs from `font-property`. I've been wanting to clean some of these (e.g. high/top/sink-priority for instance).

darinadler replied:
I think it’s tricky to use the list of longhands but I will do it if I can figure it out. If not I will return and switch to that.

darinadler replied:
OK, after a little bit of a journey learning Python, this is done.

nt1m commented on Source/WebCore/style/StyleBuilderCustom.h:
```diff
@@ -1680,6 +1689,10 @@ inline void BuilderCustom::applyInitialFontVariantNumeric(BuilderState& builderS
 inline void BuilderCustom::applyValueFontVariantNumeric(BuilderState& builderState, CSSValue& value)
 {
+    if (is<CSSPrimitiveValue>(value) && CSSPropertyParserHelpers::isSystemFontShorthand(downcast<CSSPrimitiveValue>(value).valueID())) {
+        applyInitialFontVariantLigatures(builderState);
```
s/applyInitialFontVariantLigatures/applyInitialFontVariantNumeric ?

darinadler replied:
Oh no, big mistake! Must fix this.

nt1m commented on Source/WebCore/style/StyleBuilderCustom.h:
```diff
@@ -1710,6 +1723,10 @@ inline void BuilderCustom::applyInitialFontVariantEastAsian(BuilderState& builde
 inline void BuilderCustom::applyValueFontVariantEastAsian(BuilderState& builderState, CSSValue& value)
 {
+    if (is<CSSPrimitiveValue>(value) && CSSPropertyParserHelpers::isSystemFontShorthand(downcast<CSSPrimitiveValue>(value).valueID())) {
+        applyInitialFontVariantLigatures(builderState);
```
ditto here with EastAsian

darinadler replied:
Ouch! Will fix.

nt1m commented on Source/WebCore/style/StyleBuilderConverter.h:
```diff
@@ -1362,29 +1358,14 @@ inline std::optional<FontSelectionValue> BuilderConverter::convertFontStyleFromV
-        if (valueID == CSSValueLighter)
-            return FontCascadeDescription::lighterWeight(builderState.parentStyle().fontDescription().weight());
-        if (CSSPropertyParserHelpers::isSystemFontShorthand(valueID))
-            return SystemFontDatabase::singleton().systemFontShorthandWeight(CSSPropertyParserHelpers::lowerFontShorthand(valueID));
```
Does this cause a change in behavior? I assume we still want the system font weight, one way or another. I guess we'd need custom apply functions to do properly with the new setup.

darinadler replied:
Good catch. That’s a mistake.

darinadler replied:
It’s a mess because I’d have to make both a custom apply function and a custom initial function, even though I don’t really want either. The other option is to require the `isSystemFontShorthand` in all converters, and I think that’s what I’ll do.

nt1m commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -5334,102 +5350,63 @@ bool CSSPropertyParser::consumeFont(bool important)
+        addProperty(property, CSSPropertyFont, CSSValuePool::singleton().createIdentifierValue(initialValue), important, true);
     };
     auto add = [&] (CSSPropertyID property, RefPtr<CSSValue>& value) {
         if (value)
             addProperty(property, CSSPropertyFont, value.releaseNonNull(), important);
         else
-            reset(property);
+            reset(property, CSSValueNormal);
     };
```
This `add` function is very similar to `addPropertyWithImplicitDefault`.

I think that method could be used directly, and `addProperty` could be used for `font-size`/`font-family` (since they are mandatory).

Or if you want to avoid the verbosity, you could also rewrite `add` to be:
```
auto add = [&] (CSSPropertyID property, RefPtr<CSSValue>& value) {
    if (property == CSSPropertyFontSize || property == CSSPropertyFontFamily)
        addProperty(property, CSSPropertyFont, value.releaseNonNull(), important);
    else
        addPropertyWithImplicitDefault(property, CSSPropertyFont, value.releaseNonNull(), CSSValuePool::singleton().createIdentifierValue(CSSValueNormal), important);
};
```

nt1m replied:
(could also be done separately if you prefer)

darinadler committed:
StyleProperties::fontValue needs to return emptyString in many more cases where font sub properties have unusual values
https://bugs.webkit.org/show_bug.cgi?id=247498
rdar://problem/102248342

Reviewed by Tim Nguyen and Sam Weinig.

To make the font serialization work correctly, had to change font shorthand parsing to set font
subproperties to initial values, not to the keyword "initial". To make system font serialization
work correctly, had to change font shorthand parsing to set the system font for all longhands,
not just a select few.

* LayoutTests/fast/css/font-shorthand-expected.txt: Rebased since we use initial values, not
values that are literally "initial", for the font subproperties set by the font shorthand.

* LayoutTests/TestExpectations: Removed expectations for tests that were already removed.

* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/font-shorthand-serialization-prevention-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/font-shorthand-serialization-prevention.html:
Added.

* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-shorthand-variant-expected.txt:
Expect PASS.

* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/system-fonts-serialization.tentative-expected.txt:
Expect PASS.

* LayoutTests/imported/w3c/web-platform-tests/css/cssom/font-shorthand-serialization-expected.txt:
Rebased since we use initial values, not values that are literally "initial", for the font
subproperties set by the font shorthand.
* LayoutTests/platform/win/fast/css/font-shorthand-expected.txt: Ditto.

* LayoutTests/platform/win/imported/w3c/web-platform-tests/css/css-fonts/font-shorthand-serialization-prevention-expected.txt:
Added.

* Source/WebCore/css/CSSProperties.json: Removed the converter codegen property from
font-variant-caps, since the generator now handles it correctly.

* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::fontShorthandValue): Added checking for font-kerning, and removed the FIXME.

* Source/WebCore/css/StyleProperties.cpp:
(WebCore::valueID): Added.
(WebCore::StyleProperties::propertyAsValueID const): Changed to return a std::optional so
callers can tell a missing value from a non-valueID value; useful in more places.
(WebCore::StyleProperties::isSingleFontShorthand const): Deleted.
(WebCore::StyleProperties::fontValue const): Combined the CSS-wide keyword logic with the system
font keyword logic, and make sure we don't ever try to use either as an individual value in the
font shorthand. Added checking for the font subproperties that are reset by the font shorthand.
If any are set to non-initial values, then serialize as the empty string.
(WebCore::StyleProperties::fontSynthesisValue const): Removed valueID lambda since we can use the
function added above instead.
(WebCore::StyleProperties::getPropertyCSSValue const): Simplify a bit using the valueID function.

* Source/WebCore/css/StyleProperties.h: Changed return type of propertyAsValueID to std::optional.
Removed isSingleFontShorthand.

* Source/WebCore/css/StylePropertyShorthand.h: Added array for handling the font shorthand rules
for resetting font subproperties to their initial values, named
fontShorthandSubpropertiesResetToInitialValues. Used in two places. Maybe generate in future?

* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::CSSPropertyParser::consumeFont): Rewrote system font handling so it puts the system
font value into all font subproperties, not just a select few. Changed property resetting so
it resets properties to their initial values, not to the "initial" keyword.

* Source/WebCore/css/process-css-properties.py:
(GenerationContext): Generate code to apply the initial value when we encounter a system font.

* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertFontFeatureSettings): Handle system font.
(WebCore::Style::BuilderConverter::convertFontVariantCaps): Deleted. This is now generated.
(WebCore::Style::BuilderConverter::convertFontVariationSettings): Handle system font.
(WebCore::Style::BuilderConverter::convertFontPalette): Ditto.

* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyValueLineHeight): Handle system font.
(WebCore::Style::BuilderCustom::applyValueFontVariantLigatures): Ditto.
(WebCore::Style::BuilderCustom::applyValueFontVariantNumeric): Ditto.
(WebCore::Style::BuilderCustom::applyValueFontVariantEastAsian): Ditto.
(WebCore::Style::BuilderCustom::applyValueFontVariantAlternates): Ditto.
(WebCore::Style::BuilderCustom::applyValueFontSizeAdjust): Ditto.

* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::getPlatformEditorStateCommon const): Updated for change to propertyAsValueID.

Canonical link: https://commits.webkit.org/256635@main
